submitit/slurm/slurm.py

```diff
@@ -140,6 +145,66 @@ def _requeue(self, countdown: int) -> None:
         subprocess.check_call(["scontrol", "requeue", jid])
         logger.get_logger().info(f"Requeued job {jid} ({countdown} remaining timeouts)")
 
+    def _parse_nodelist(self, node_list: str) -> List[str]:
+        """Parse the content of the "nodes" environment variable,
+        which gives access to the list of hostnames that are part
+        of the current job.
+
+        In SLURM, the node list is formatted NODE_GROUP_1,NODE_GROUP_2,...,NODE_GROUP_N
+        where each node group is formatted as: PREFIX[1-3,5,8] to define the hosts:
+        [PREFIX1, PREFIX2, PREFIX3, PREFIX5, PREFIX8].
+
+        Link: https://hpcc.umd.edu/hpcc/help/slurmenv.html
+        """
+        if not node_list:
+            return [self.hostname]
+
+        try:
+            pos = 0
+            parsed: List[str] = []
+            while pos < len(node_list):
+                pos = self._parse_group(node_list, pos, parsed)
+            return parsed
+        except ValueError as e:
+            raise SlurmParseException(f"Unrecognized format for SLURM_JOB_NODELIST: '{node_list}'", e)
+
+    def _parse_group(self, node_list: str, pos: int, parsed: List[str]) -> int:
+        """Parse a suffix group of the form PREFIX[1-3,5,8] and return
+        the position in the string at which the parsing stopped
+        """
+        prefixes = [""]
+        while pos < len(node_list):
+            c = node_list[pos]
+            if c == ",":
+                parsed.extend(prefixes)
+                return pos + 1
+            if c == "[":
+                last_pos = node_list.index("]", pos)
+                suffixes = self._expand_suffix(node_list[pos + 1 : last_pos])
+                prefixes = [prefix + suffix for prefix in prefixes for suffix in suffixes]
+                pos = last_pos + 1
+            else:
+                for i, prefix in enumerate(prefixes):
+                    prefixes[i] = prefix + c
+                pos += 1
+        parsed.extend(prefixes)
+        return pos
+
+    @staticmethod
+    def _expand_suffix(suffix_parts: str) -> List[str]:
```

Comment:
Private static method *in my opinion* are useless. In my view (but thats arguable), static methods are only useful to provide some functions through a namespace for the users, otherwise it's simpler to have an actual private function

Reply:
Agree with Jérémy on this: I'd rather see the two methods buried at the end of the file than inside the SlurmJobEnvironment class.